File: python/minesweeper/minesweeper.py
def annotate(minefield):
    minefield = [list(row) for row in minefield]

    for row in range(len(minefield)):
        if len(minefield[row]) != len(minefield[0]):
            raise ValueError(r".+")

        for col in range(len(minefield[0])):
            if minefield[row][col] == ' ':
                counter = 0

                for r in range(max(row - 1, 0), min(row + 2, len(minefield))):
                    for c in range(max(col - 1, 0), min(col + 2, len(minefield[0]))):
                        if minefield[r][c] == '*':
                            counter += 1

                if counter > 0:
                    minefield[row][col] = str(counter)
            elif minefield[row][col] != '*':
                raise ValueError(r".+")

    return [''.join(row) for row in minefield]

File: haskell/minesweeper/src/Minesweeper.hs
module Minesweeper
  ( annotate
  ) where

import           Data.Char                      ( intToDigit )

annotate :: [String] -> [String]
annotate []    = []
annotate board = foldr
  f
  board
  [ (r, c) | r <- [0 .. rows - 1], c <- [0 .. cols - 1] ]
 where
  rows = length board
  cols = length (head board)
  f (i, j) xs
    | xs !! i !! j == ' ' && mines > 0 = setAt i j (intToDigit mines) xs
    | otherwise                        = xs
   where
    mines = sum
      [ 1
      | i' <- [max (i - 1) 0 .. min (i + 1) (rows - 1)]
      , j' <- [max (j - 1) 0 .. min (j + 1) (cols - 1)]
      , xs !! i' !! j' == '*'
      ]
  setAt i j x xs = take i xs ++ [setAt' j x (xs !! i)] ++ drop (i + 1) xs
  setAt' j x xs = take j xs ++ [x] ++ drop (j + 1) xs
